### PR TITLE
Feature camera info publisher

### PR DESCRIPTION
**Pull request by TheMangalex**
Ros camera sensors normally publish a camera info next to image topics. This pull request aims to provide a basic version for the Camera and RangeFinder Ros Controller.

### CODE REVIEW

File: projects/default/controllers/ros/RosCamera.cpp [hunk 1]
```diff
@@ -76,9 +76,15 @@ RosCamera::~RosCamera() {
 // creates a publisher for camera image with
 // a [4 x ImageWidth x ImageHeight] {unsigned char} array
 ros::Publisher RosCamera::createPublisher() {
+  createCameraInfoPublisher("camera_info");
```
Comment (omichel) [inline projects/default/controllers/ros/RosCamera.cpp:79]: This parameter is always set to "camera_info", therefore I believe it should be hard-coded in the `createCameraInfoPublisher` method:
```suggestion
  createCameraInfoPublisher();
```

File: projects/default/controllers/ros/RosCamera.cpp [hunk 2]
```diff
@@ -76,9 +76,15 @@ RosCamera::~RosCamera() {
...
   return createImagePublisher("image");
 }
 
+void RosCamera::createCameraInfoPublisher(const std::string &name) {
+  sensor_msgs::CameraInfo type;
+  mCameraInfoPublisher = RosDevice::rosAdvertiseTopic(RosDevice::fixedDeviceName() + "/" + name, type);
```
Comment (omichel) [inline projects/default/controllers/ros/RosCamera.cpp:85]:
```suggestion
void RosCamera::createCameraInfoPublisher() {
  sensor_msgs::CameraInfo type;
  mCameraInfoPublisher = RosDevice::rosAdvertiseTopic(RosDevice::fixedDeviceName() + "/camera_info", type);
```

File: projects/default/controllers/ros/RosCamera.cpp [hunk 3]
```diff
@@ -107,7 +113,41 @@ void RosCamera::publishValue(ros::Publisher publisher) {
...
 
+sensor_msgs::CameraInfo RosCamera::createCameraInfoMessage() {
+  sensor_msgs::CameraInfo info;
+  info.header.stamp = ros::Time::now();
+  info.header.frame_id = mFrameIdPrefix + RosDevice::fixedDeviceName();
+
+  double width = mCamera->getWidth();
+  double height = mCamera->getHeight();
+  info.width = mCamera->getWidth();
+  info.height = mCamera->getHeight();
```
Comment (omichel) [inline projects/default/controllers/ros/RosCamera.cpp:124]: Cleaner this way:
```suggestion
  const double width = mCamera->getWidth();
  const double height = mCamera->getHeight();
  info.width = width;
  info.height = height;
```

File: projects/default/controllers/ros/RosCamera.cpp [hunk 4]
```diff
@@ -107,7 +113,41 @@ void RosCamera::publishValue(ros::Publisher publisher) {
...
+
+  double hfov = mCamera->getFov();
+  double focalLength = width / (2.0 * tan(hfov / 2.0));
```
Comment (omichel) [inline projects/default/controllers/ros/RosCamera.cpp:127]:
```suggestion
  double horizontalFov = mCamera->getFov();
  double focalLength = width / (2.0 * tan(horizontalFov / 2.0));
```

File: projects/default/controllers/ros/RosCamera.cpp [hunk 5]
```diff
@@ -107,7 +113,41 @@ void RosCamera::publishValue(ros::Publisher publisher) {
...
+
+  double fx, fy, cx, cy;
+  fx = focalLength;
+  fy = focalLength;
+  cx = (width + 1.0) / 2.0;
+  cy = (height + 1.0) / 2.0;
```
Comment (omichel) [inline projects/default/controllers/ros/RosCamera.cpp:133]:
```suggestion
  const double fx = focalLength;
  const double fy = focalLength;
  const double cx = (width + 1.0) / 2.0;
  const double cy = (height + 1.0) / 2.0;
```

File: projects/default/controllers/ros/RosCamera.cpp [hunk 6]
```diff
@@ -107,7 +113,41 @@ void RosCamera::publishValue(ros::Publisher publisher) {
...
+  boost::array<double, 9> K = {fx, 0.0, cx, 0.0, fy, cy, 0.0, 0.0, 1.0};
```
Comment (omichel) [inline projects/default/controllers/ros/RosCamera.cpp:134]:
```suggestion
  const boost::array<double, 9> K = {fx, 0.0, cx, 0.0, fy, cy, 0.0, 0.0, 1.0};
```

File: projects/default/controllers/ros/RosCamera.cpp [hunk 7]
```diff
@@ -107,7 +113,41 @@ void RosCamera::publishValue(ros::Publisher publisher) {
...
+  info.K = K;
+
+  boost::array<double, 9> R = {1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0};
```
Comment (omichel) [inline projects/default/controllers/ros/RosCamera.cpp:137]:
```suggestion
  const boost::array<double, 9> R = {1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0};
```

File: projects/default/controllers/ros/RosCamera.cpp [hunk 8]
```diff
@@ -107,7 +113,41 @@ void RosCamera::publishValue(ros::Publisher publisher) {
...
+  info.R = R;
+
+  boost::array<double, 12> P = {fx, 0.0, cx, 0.0, 0.0, fy, cy, 0.0, 0.0, 0.0, 1.0, 0.0};
```
Comment (omichel) [inline projects/default/controllers/ros/RosCamera.cpp:140]:
```suggestion
  const boost::array<double, 12> P = {fx, 0.0, cx, 0.0, 0.0, fy, cy, 0.0, 0.0, 0.0, 1.0, 0.0};
```

File: projects/default/controllers/ros/RosCamera.cpp [hunk 9]
```diff
@@ -107,7 +113,41 @@ void RosCamera::publishValue(ros::Publisher publisher) {
...
+  info.P = P;
+
+  return info;
+}
+
 void RosCamera::publishAuxiliaryValue() {
+  if (mCameraInfoPublisher.getNumSubscribers() > 0) {
+    mCameraInfoPublisher.publish(createCameraInfoMessage());
+  }
```
Comment (omichel) [inline projects/default/controllers/ros/RosCamera.cpp:149]: Coding style [#CS360](https://github.com/cyberbotics/webots/wiki/CPP-Coding-Style#cs360-single-if-while-or-for-statement-must-be-written-without-brackets-the-statement-part-should-be-put-on-a-separate-line-in-case-of-nested-block-keep-brackets):
```suggestion
  if (mCameraInfoPublisher.getNumSubscribers() > 0)
    mCameraInfoPublisher.publish(createCameraInfoMessage());
```

File: projects/default/controllers/ros/RosCamera.hpp [hunk 10]
```diff
@@ -61,12 +62,15 @@ class RosCamera : public RosSensor {
 private:
   ros::Publisher createImagePublisher(const std::string &name);
   void cleanup() { mCamera->disable(); }
+  void createCameraInfoPublisher(const std::string &name);
```
Comment (omichel) [inline projects/default/controllers/ros/RosCamera.hpp:65]:
```suggestion
  void createCameraInfoPublisher();
```

**Review by omichel — CHANGES_REQUESTED** [review]
Thank you for this nice contribution.
I only have a few requests about coding style.
May I let you apply my comments also to the range finder class?

File: projects/default/controllers/ros/RosCamera.cpp (hunk 1 above)
Comment (TheMangalex) [reply projects/default/controllers/ros/RosCamera.cpp:79]: Hi, I can adept this to a fixed name and update the pull request.

Comment (omichel) [reply projects/default/controllers/ros/RosCamera.cpp]: That would be great.

Comment (TheMangalex) [reply projects/default/controllers/ros/RosCamera.cpp]: Changes should be done in the newest Push. Feel free to confirm, if I did everything correct. ^^

File: projects/default/controllers/ros/RosCamera.cpp [hunk 11]
```diff
@@ -107,7 +113,40 @@ void RosCamera::publishValue(ros::Publisher publisher) {
...
+  info.header.stamp = ros::Time::now();
+  info.header.frame_id = mFrameIdPrefix + RosDevice::fixedDeviceName();
+
+  const double width = mCamera->getWidth();
+  const double height = mCamera->getHeight();
+  info.width = width;
+  info.height = height;
+
+  const double horizontalFov = mCamera->getFov();
+  double focalLength = width / (2.0 * tan(horizontalFov / 2.0));
```
Comment (omichel) [inline projects/default/controllers/ros/RosCamera.cpp:127]: I believe this one could be const as well.
```suggestion
  const double focalLength = width / (2.0 * tan(horizontalFov / 2.0));
```

File: projects/default/controllers/ros/RosRangeFinder.cpp [hunk 12]
```diff
@@ -60,6 +66,41 @@ void RosRangeFinder::publishValue(ros::Publisher publisher) {
...
+  info.header.stamp = ros::Time::now();
+  info.header.frame_id = mFrameIdPrefix + RosDevice::fixedDeviceName();
+
+  const double width = mRangeFinder->getWidth();
+  const double height = mRangeFinder->getHeight();
+  info.width = width;
+  info.height = height;
+
+  const double horizontalFov = mRangeFinder->getFov();
+  double focalLength = width / (2.0 * tan(horizontalFov / 2.0));
```
Comment (omichel) [inline projects/default/controllers/ros/RosRangeFinder.cpp:80]:
```suggestion
  const double focalLength = width / (2.0 * tan(horizontalFov / 2.0));
```

**Review by omichel — CHANGES_REQUESTED** [review]
It looks all good to me except for one missing `const`.

**Comment by TheMangalex** [comment]
Seems like that slipped through. Should be done now.

**Review by omichel — APPROVED** [review]
Very good. Thank you.

**Pull request merged**